Added metrics cardinality handling for Google Cloud Spanner receiver - part 3.

ydrozhdzhal opened the pull request:
Final part to bind all together: algorithm, filter resolver metrics builder, receiver and config.

Commit:
Added metrics cardinality handling for Google Cloud Spanner receiver - part 3.

Final part to bind all together: algorithm, filter resolver metrics builder, receiver and config.

ydrozhdzhal commented:
@MovieStoreGuy , @jpkrohling this is the final, 3d part.

MovieStoreGuy commented on receiver/googlecloudspannerreceiver/config_test.go:
```diff
@@ -27,6 +27,10 @@ import (
 )
 
+const (
+	cardinalityOverallLimit = 200000
```
```suggestion
cardinalityLimit = 200000
```

keeping the naming simple :)

ydrozhdzhal replied:
Done.

MovieStoreGuy reviewed (commented):
So far it looks fine, be happy to approve once testify mock has been adopted, just simplifies the defined types.

ydrozhdzhal commented:
@MovieStoreGuy , looks like that in this particular case using of mocks makes tests more complicated and it is hard to adopt.

MovieStoreGuy reviewed (approved):
This looks solid.

ydrozhdzhal commented:
@jpkrohling , @bogdandrutu , all reviews received, except maintainer's one. Can you look at this and approve and merge if everything is OK?

jpkrohling commented on receiver/googlecloudspannerreceiver/README.md:
```diff
@@ -52,9 +53,10 @@ receivers:
 Brief description of configuration properties:
 - **googlecloudspanner** - name of the Cloud Spanner Receiver related section in OpenTelemetry collector configuration file
-- **collection_interval** - this receiver runs on an interval. Each time it runs, it queries Google Cloud Spanner, creates metrics and sends them to the next consumer(60 seconds by default)
+- **collection_interval** - this receiver runs on an interval. Each time it runs, it queries Google Cloud Spanner, creates metrics and sends them to the next consumer(60 seconds by default). **It is not recommended change the default value fo collection interval, since new values for metrics in Spanner database appear with 1 minute interval.**
```
```suggestion
- **collection_interval** - this receiver runs periodically. Each time it runs, it queries Google Cloud Spanner, creates metrics, and sends them to the next consumer (default: 1 minute). **It is not recommended to change the default value of collection interval, since new values for metrics in the Spanner database appear only once a minute.**
```

ydrozhdzhal replied:
Done.

jpkrohling commented on receiver/googlecloudspannerreceiver/README.md:
```diff
@@ -52,9 +53,10 @@ receivers:
+- **collection_interval** - this receiver runs on an interval. Each time it runs, it queries Google Cloud Spanner, creates metrics and sends them to the next consumer(60 seconds by default). **It is not recommended change the default value fo collection interval, since new values for metrics in Spanner database appear with 1 minute interval.**
 - **top_metrics_query_max_rows** - max number of rows to fetch from Top N built-in table(100 by default)
 - **backfill_enabled** - turn on/off 1-hour data backfill(by default it is turned off)
+- **cardinality_total_limit** - total cardinality handling limit(amount of active series per 24 hours period). If specified - turns on cardinality filtering/handling. If zero or not specified - cardinality is not handled. You can read [this document](cardinality.md) for getting more information about cardinality handling/filtering.
```
```suggestion
- **cardinality_total_limit** - limit of active series per 24 hours period. If specified, turns on cardinality filtering and handling. If zero or not specified, cardinality is not handled. You can read [this document](cardinality.md) for more information about cardinality handling and filtering.
```

ydrozhdzhal replied:
Done.

jpkrohling commented on receiver/googlecloudspannerreceiver/config.go:
```diff
@@ -55,6 +56,11 @@ func (config *Config) Validate() error {
 	}
 
+	if config.CardinalityTotalLimit < 0 {
+		return fmt.Errorf("%v `cardinality_total_limit` must be not negative, current value is %vs",
```
What's this `s` at the end of the format? Also, try to use double quotes instead of backticks here.

ydrozhdzhal replied:
Done. `s` - it was my mistake ((

jpkrohling commented on receiver/googlecloudspannerreceiver/config_test.go:
```diff
@@ -27,6 +27,10 @@ import (
 )
 
+const (
+	cardinalityLimit = 200000
```
Tip: you can use underscores to make the number more readable.

```suggestion
	cardinalityLimit = 200_000
```

ydrozhdzhal replied:
Done.

jpkrohling reviewed (commented):
Looks good, just a couple of small changes to the docs.

ydrozhdzhal commented:
@jpkrohling , addressed all comments, re-requested review. In case of merging this I will ask you the following: when squashing commits - can you leave only the main commit message and do not include messages from CR related commits? Other option is to allow me squash before merging.

ydrozhdzhal commented:
Re-based by myself, to re-trigger flaky test jobs.

jpkrohling commented:
Please, do not squash the commits between reviews, as it makes it harder to spot the differences between a code that has been reviewed already and the newest changes.

jpkrohling commented on receiver/googlecloudspannerreceiver/config.go:
```diff
@@ -42,21 +43,27 @@ type Instance struct {
 	if config.CollectionInterval.Seconds() < 60 {
-		return fmt.Errorf("%v `collection_interval` must be not lower than 60s, current value is %vs",
-			config.ID(), config.CollectionInterval.Seconds())
+		return fmt.Errorf("%v %q must be not lower than 60s, current value is %vs", config.ID(),
```
```suggestion
		return fmt.Errorf("%v %q must be not lower than 60s, current value is %v", config.ID(),
```

ydrozhdzhal replied:
In this particular case s means seconds

jpkrohling replied:
Got it. Thanks for the confirmation.
